File: 2023/day24/c/solution.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>

#define MAX_HAILSTONES 400

typedef long long i64;
typedef long double ld;

typedef struct {
    i64 px, py, pz;
    i64 vx, vy, vz;
} Hailstone;

// Parse hailstones from file
static int parse_input(const char *filename, Hailstone *hailstones) {
    FILE *f = fopen(filename, "r");
    if (!f) {
        perror("fopen");
        exit(1);
    }

    int count = 0;
    char line[256];
    while (fgets(line, sizeof(line), f)) {
        Hailstone *h = &hailstones[count];
        if (sscanf(line, "%lld, %lld, %lld @ %lld, %lld, %lld",
                   &h->px, &h->py, &h->pz, &h->vx, &h->vy, &h->vz) == 6) {
            count++;
        }
    }
    fclose(f);
    return count;
}

// Part 1: Find 2D intersections in future within bounds
static int part1(Hailstone *h, int n) {
    const double MIN_COORD = 200000000000000.0;
    const double MAX_COORD = 400000000000000.0;
    int count = 0;

    for (int i = 0; i < n; i++) {
        for (int j = i + 1; j < n; j++) {
            double vx1 = h[i].vx, vy1 = h[i].vy;
            double vx2 = h[j].vx, vy2 = h[j].vy;
            double px1 = h[i].px, py1 = h[i].py;
            double px2 = h[j].px, py2 = h[j].py;

            double det = vx1 * (-vy2) - (-vx2) * vy1;
            if (det == 0) continue;

            double dx = px2 - px1;
            double dy = py2 - py1;

            double t1 = (dx * (-vy2) - (-vx2) * dy) / det;
            double t2 = (vx1 * dy - dx * vy1) / det;

            if (t1 < 0 || t2 < 0) continue;

            double x = px1 + vx1 * t1;
            double y = py1 + vy1 * t1;

            if (x >= MIN_COORD && x <= MAX_COORD &&
                y >= MIN_COORD && y <= MAX_COORD) {
                count++;
            }
        }
    }

    return count;
}

// Solve 4x4 linear system using Gaussian elimination with long double
static void solve_system_4x4_ld(ld matrix[4][5], ld solution[4]) {
    // Forward elimination with partial pivoting
    for (int col = 0; col < 4; col++) {
        // Find pivot
        int max_row = col;
        for (int row = col + 1; row < 4; row++) {
            if (fabsl(matrix[row][col]) > fabsl(matrix[max_row][col])) {
                max_row = row;
            }
        }

        // Swap rows
        for (int k = 0; k < 5; k++) {
            ld tmp = matrix[col][k];
            matrix[col][k] = matrix[max_row][k];
            matrix[max_row][k] = tmp;
        }

        if (matrix[col][col] == 0) continue;

        // Eliminate column
        for (int row = col + 1; row < 4; row++) {
            if (matrix[row][col] != 0) {
                ld factor = matrix[row][col] / matrix[col][col];
                for (int k = col; k < 5; k++) {
                    matrix[row][k] -= factor * matrix[col][k];
                }
            }
        }
    }

    // Back substitution
    for (int i = 3; i >= 0; i--) {
        solution[i] = matrix[i][4];
        for (int j = i + 1; j < 4; j++) {
            solution[i] -= matrix[i][j] * solution[j];
        }
        solution[i] /= matrix[i][i];
    }
}

// Part 2: Find rock position that hits all hailstones
static i64 part2(Hailstone *h, int n __attribute__((unused))) {
    // Build system for XY plane (4 equations, 4 unknowns: rx, ry, rvx, rvy)
    // For hailstones i and j:
    // (vyi - vyj)*rx + (vxj - vxi)*ry + (pyj - pyi)*rvx + (pxi - pxj)*rvy
    //   = pxi*vyi - pyi*vxi - (pxj*vyj - pyj*vxj)

    ld matrix_xy[4][5];
    for (int i = 0; i < 4; i++) {
        ld px1 = h[i].px, py1 = h[i].py;
        ld vx1 = h[i].vx, vy1 = h[i].vy;
        ld px2 = h[i+1].px, py2 = h[i+1].py;
        ld vx2 = h[i+1].vx, vy2 = h[i+1].vy;

        matrix_xy[i][0] = vy1 - vy2;
        matrix_xy[i][1] = vx2 - vx1;
        matrix_xy[i][2] = py2 - py1;
        matrix_xy[i][3] = px1 - px2;
        matrix_xy[i][4] = px1*vy1 - py1*vx1 - (px2*vy2 - py2*vx2);
    }

    ld sol_xy[4];
    solve_system_4x4_ld(matrix_xy, sol_xy);
    i64 rx = (i64)roundl(sol_xy[0]);
    i64 ry = (i64)roundl(sol_xy[1]);

    // Build system for XZ plane to get rz
    ld matrix_xz[4][5];
    for (int i = 0; i < 4; i++) {
        ld px1 = h[i].px, pz1 = h[i].pz;
        ld vx1 = h[i].vx, vz1 = h[i].vz;
        ld px2 = h[i+1].px, pz2 = h[i+1].pz;
        ld vx2 = h[i+1].vx, vz2 = h[i+1].vz;

        matrix_xz[i][0] = vz1 - vz2;
        matrix_xz[i][1] = vx2 - vx1;
        matrix_xz[i][2] = pz2 - pz1;
        matrix_xz[i][3] = px1 - px2;
        matrix_xz[i][4] = px1*vz1 - pz1*vx1 - (px2*vz2 - pz2*vx2);
    }

    ld sol_xz[4];
    solve_system_4x4_ld(matrix_xz, sol_xz);
    i64 rz = (i64)roundl(sol_xz[1]);  // Second element is rz

    return rx + ry + rz;
}

int main(int argc, char **argv) {
    const char *filename = argc > 1 ? argv[1] : "../input.txt";

    Hailstone hailstones[MAX_HAILSTONES];
    int n = parse_input(filename, hailstones);

    printf("Part 1: %d\n", part1(hailstones, n));
    printf("Part 2: %lld\n", part2(hailstones, n));

    return 0;
}

File: 2023/day24/rust/solution.rs
use std::fs;

#[derive(Clone, Copy, Debug)]
struct Hailstone {
    px: i64,
    py: i64,
    pz: i64,
    vx: i64,
    vy: i64,
    vz: i64,
}

fn parse_input(filename: &str) -> Vec<Hailstone> {
    let content = fs::read_to_string(filename).expect("Failed to read input file");
    content
        .lines()
        .filter(|line| !line.is_empty())
        .map(|line| {
            let parts: Vec<&str> = line.split('@').collect();
            let pos: Vec<i64> = parts[0]
                .split(',')
                .map(|s| s.trim().parse().unwrap())
                .collect();
            let vel: Vec<i64> = parts[1]
                .split(',')
                .map(|s| s.trim().parse().unwrap())
                .collect();
            Hailstone {
                px: pos[0],
                py: pos[1],
                pz: pos[2],
                vx: vel[0],
                vy: vel[1],
                vz: vel[2],
            }
        })
        .collect()
}

/// Find intersection of two hailstone paths in 2D (XY plane).
/// Returns Some((x, y, t1, t2)) or None if parallel.
fn find_intersection_2d(h1: &Hailstone, h2: &Hailstone) -> Option<(f64, f64, f64, f64)> {
    let vx1 = h1.vx as f64;
    let vy1 = h1.vy as f64;
    let vx2 = h2.vx as f64;
    let vy2 = h2.vy as f64;
    let px1 = h1.px as f64;
    let py1 = h1.py as f64;
    let px2 = h2.px as f64;
    let py2 = h2.py as f64;

    // Solve for t1 and t2 using Cramer's rule:
    // vx1*t1 - vx2*t2 = px2 - px1
    // vy1*t1 - vy2*t2 = py2 - py1
    let det = vx1 * (-vy2) - (-vx2) * vy1;
    if det.abs() < 1e-10 {
        return None; // Parallel lines
    }

    let dx = px2 - px1;
    let dy = py2 - py1;

    let t1 = (dx * (-vy2) - (-vx2) * dy) / det;
    let t2 = (vx1 * dy - dx * vy1) / det;

    let x = px1 + vx1 * t1;
    let y = py1 + vy1 * t1;

    Some((x, y, t1, t2))
}

fn part1(hailstones: &[Hailstone]) -> usize {
    const MIN_COORD: f64 = 200000000000000.0;
    const MAX_COORD: f64 = 400000000000000.0;
    let mut count = 0;

    for i in 0..hailstones.len() {
        for j in (i + 1)..hailstones.len() {
            if let Some((x, y, t1, t2)) = find_intersection_2d(&hailstones[i], &hailstones[j]) {
                // Check if intersection is in the future for both hailstones
                if t1 < 0.0 || t2 < 0.0 {
                    continue;
                }

                // Check if intersection is within test area
                if x >= MIN_COORD && x <= MAX_COORD && y >= MIN_COORD && y <= MAX_COORD {
                    count += 1;
                }
            }
        }
    }

    count
}

/// Solve a system of linear equations using Gaussian elimination with f64.
/// Uses partial pivoting for numerical stability.
fn solve_system(matrix: &[[f64; 4]; 4], rhs: &[f64; 4]) -> [f64; 4] {
    const N: usize = 4;
    // Create augmented matrix
    let mut aug: [[f64; 5]; 4] = [[0.0; 5]; 4];
    for i in 0..N {
        for j in 0..N {
            aug[i][j] = matrix[i][j];
        }
        aug[i][N] = rhs[i];
    }

    // Forward elimination with partial pivoting
    for col in 0..N {
        // Find pivot (row with largest absolute value in this column)
        let mut max_row = col;
        for row in (col + 1)..N {
            if aug[row][col].abs() > aug[max_row][col].abs() {
                max_row = row;
            }
        }
        aug.swap(col, max_row);

        if aug[col][col].abs() < 1e-15 {
            continue;
        }

        // Eliminate column below pivot
        for row in (col + 1)..N {
            let factor = aug[row][col] / aug[col][col];
            for j in col..=N {
                aug[row][j] -= factor * aug[col][j];
            }
        }
    }

    // Back substitution
    let mut solution = [0.0; 4];
    for i in (0..N).rev() {
        solution[i] = aug[i][N];
        for j in (i + 1)..N {
            solution[i] -= aug[i][j] * solution[j];
        }
        solution[i] /= aug[i][i];
    }

    solution
}

fn part2(hailstones: &[Hailstone]) -> i64 {
    // Build system for XY plane (4 equations, 4 unknowns: rx, ry, rvx, rvy)
    //
    // The key insight is that for each hailstone i, if the rock hits it at time ti:
    // rx + rvx*ti = pxi + vxi*ti
    // ry + rvy*ti = pyi + vyi*ti
    //
    // Eliminating ti gives:
    // (rx - pxi) * (vyi - rvy) = (ry - pyi) * (vxi - rvx)
    //
    // Expanding: rx*vyi - rx*rvy - pxi*vyi + pxi*rvy = ry*vxi - ry*rvx - pyi*vxi + pyi*rvx
    //
    // The terms rx*rvy and ry*rvx are nonlinear. However, if we take two hailstones i and j
    // and subtract their equations, these nonlinear terms cancel:
    //
    // (vyi - vyj)*rx + (vxj - vxi)*ry + (pyj - pyi)*rvx + (pxi - pxj)*rvy
    //     = pxi*vyi - pyi*vxi - (pxj*vyj - pyj*vxj)
    //
    // This gives us linear equations! We need 4 equations (from 5 hailstones) for 4 unknowns.

    let h = &hailstones[..5];

    let mut matrix_xy: [[f64; 4]; 4] = [[0.0; 4]; 4];
    let mut rhs_xy: [f64; 4] = [0.0; 4];

    for i in 0..4 {
        let px1 = h[i].px as f64;
        let py1 = h[i].py as f64;
        let vx1 = h[i].vx as f64;
        let vy1 = h[i].vy as f64;
        let px2 = h[i + 1].px as f64;
        let py2 = h[i + 1].py as f64;
        let vx2 = h[i + 1].vx as f64;
        let vy2 = h[i + 1].vy as f64;

        // Coefficients for rx, ry, rvx, rvy
        matrix_xy[i] = [vy1 - vy2, vx2 - vx1, py2 - py1, px1 - px2];
        rhs_xy[i] = px1 * vy1 - py1 * vx1 - (px2 * vy2 - py2 * vx2);
    }

    let solution_xy = solve_system(&matrix_xy, &rhs_xy);
    let rx = solution_xy[0];
    let ry = solution_xy[1];

    // Build system for XZ plane to get rz
    let mut matrix_xz: [[f64; 4]; 4] = [[0.0; 4]; 4];
    let mut rhs_xz: [f64; 4] = [0.0; 4];

    for i in 0..4 {
        let px1 = h[i].px as f64;
        let pz1 = h[i].pz as f64;
        let vx1 = h[i].vx as f64;
        let vz1 = h[i].vz as f64;
        let px2 = h[i + 1].px as f64;
        let pz2 = h[i + 1].pz as f64;
        let vx2 = h[i + 1].vx as f64;
        let vz2 = h[i + 1].vz as f64;

        // Same structure as XY but with Z instead of Y
        matrix_xz[i] = [vz1 - vz2, vx2 - vx1, pz2 - pz1, px1 - px2];
        rhs_xz[i] = px1 * vz1 - pz1 * vx1 - (px2 * vz2 - pz2 * vx2);
    }

    let solution_xz = solve_system(&matrix_xz, &rhs_xz);
    let rz = solution_xz[1];

    // Round to nearest integer (the solution should be exact integers)
    (rx.round() + ry.round() + rz.round()) as i64
}

fn main() {
    let input_file = std::env::args()
        .nth(1)
        .unwrap_or_else(|| "../input.txt".to_string());
    let hailstones = parse_input(&input_file);

    println!("Part 1: {}", part1(&hailstones));
    println!("Part 2: {}", part2(&hailstones));
}
